src/Fixer/PhpUnit/PhpUnitDataProviderMethodOrderFixer.php

```diff
@@ -0,0 +1,237 @@
+    public function dataProvider() {}
+}
+',
+                    [
+                        'placement' => 'before',
+                    ]
+                ),
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Must run before ClassAttributesSeparationFixer, NoBlankLinesAfterClassOpeningFixer.
+     * Must run after OrderedClassElementsFixer.
+     */
+    public function getPriority(): int
+    {
+        return 64;
+    }
+
+    protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
+    {
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('placement', 'Prefix that replaces "test".'))
+                ->setAllowedValues(['after', 'before'])
+                ->setDefault('after')
+                ->getOption(),
+        ]);
+    }
+
+    protected function applyPhpUnitClassFix(Tokens $tokens, int $startIndex, int $endIndex): void
+    {
+        $dataProvidersWithUsagePairs = $this->getDataProvidersWithUsagePairs($tokens, $startIndex, $endIndex);
+        $origMethodsOrderPairs = $this->getOrigMethodsOrderPairs($dataProvidersWithUsagePairs);
+        $origUsageDataProviderOrderPairs = $this->getOrigUsageDataProviderOrderPairs($dataProvidersWithUsagePairs);
+
+        $newMethodsOrder = [];
+        if ('before' === $this->configuration['placement']) {
+            foreach ($origUsageDataProviderOrderPairs as [$usageName, $providerName]) {
+                if (!isset($newMethodsOrder[$providerName])) {
+                    $newMethodsOrder[$providerName] = true;
+                }
+
+                if (!isset($newMethodsOrder[$usageName])) {
+                    $newMethodsOrder[$usageName] = true;
+                }
+            }
+        } else {
+            $providerNamesAlreadyAfter = [];
+            foreach ($origUsageDataProviderOrderPairs as [$usageName, $providerName]) {
+                if (!isset($newMethodsOrder[$usageName])) {
+                    $newMethodsOrder[$usageName] = true;
+                }
+
+                \assert(isset($origMethodsOrderPairs[$usageName], $origMethodsOrderPairs[$providerName]));
+                $isProviderAlreadyAfter = $origMethodsOrderPairs[$usageName][0] < $origMethodsOrderPairs[$providerName][0];
+
+                if (isset($newMethodsOrder[$providerName])) {
+                    if (!($providerNamesAlreadyAfter[$providerName] ?? false) || $isProviderAlreadyAfter) {
+                        unset($newMethodsOrder[$providerName]);
+                    }
+                } else {
+                    $providerNamesAlreadyAfter[$providerName] = $isProviderAlreadyAfter;
+                }
+                $newMethodsOrder[$providerName] = true;
+            }
+        }
+        $newMethodsOrder = array_keys($newMethodsOrder);
+
+        if ($newMethodsOrder === array_keys($origMethodsOrderPairs)) {
+            return;
+        }
+
+        $methodOrderFixer = new OrderedClassElementsFixer();
```

Comment:
I want to be self critical here. The PR is done, but.

The current code using `OrderedClassElementsFixer` fixer here works perfectly to assert the expected order, but the diff is not ideal as other non-test/provider methods after/before might be moved too.

This can be verified when reverting #8302 and applying this fixer.

My plan here is to make the needed `OrderedClassElementsFixer` methods public temporary and then discuss refactoring to make the required methods public. This is needed in order to not copy code.

Reply:
Done!